apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/Numeric.cs

```diff
@@ -495,9 +477,71 @@ private string ToString(int maxPrecision, LossOfPrecisionHandling lossOfPrecisio
         /// <param name="lossOfPrecisionHandling">How to handle values with signficant digits that would be lost by the conversion.</param>
         /// <exception cref="OverflowException">This value is outside the range of <see cref="decimal"/>.</exception>
         /// <returns>The converted value.</returns>
-        public decimal ToDecimal(LossOfPrecisionHandling lossOfPrecisionHandling) =>
+        public decimal ToDecimal(LossOfPrecisionHandling lossOfPrecisionHandling)
+        {
+            if (_scaledValue == BigInteger.Zero)
+            {
+                return 0m;
+            }
+            
             // Handles overflow
-            decimal.Parse(ToString(29, lossOfPrecisionHandling), CultureInfo.InvariantCulture);
+            return decimal.Parse(ToString(DecimalPrecision), CultureInfo.InvariantCulture);
+
+            int DecimalPrecision(string scaledText, int integerPlaces, int decimalPlaces)
+            {
+                int decimalPrecision;
+                // If the number had no decimal places to begin with, then we need not to worry.
+                // If the number has 29 or more integer places we cannot include any decimal places;
+                // this may overflow when we convert to decimal but that's fine, let's not handle that here.
+                if (decimalPlaces == 0 || integerPlaces >= 29)
+                {
+                    decimalPrecision = 0;
+                }
+                // If the number has at most 28 significant digits we can represent it fully.
+                else if (integerPlaces + decimalPlaces <= 28)
+                {
+                    decimalPrecision = decimalPlaces;
+                }
+                // If there are no integer places we can represent up to 28 decimal places.
+                else if (integerPlaces == 0)
+                {
+                    decimalPrecision = Math.Min(decimalPlaces, 28);
+                }
+                // integerPlaces + decimalPlaces >= 29 and
+                // 1 <= intergerPlaces <= 28 which means we may have room for some of the decimal places
+                // but we need to find out if the number can have 28 or 29 significant digits.
+                else
+                {
+                    int significantPlaces = 29;
+                    string maxDecimalText = decimal.MaxValue.ToString(CultureInfo.InvariantCulture);
```

Comment:
Do we need to do this textually? I'm wondering if we could compare with a `BigInteger` instead - presumably we *have* the integer representation at this point?